Add support for using caffe2::ThreadPool in pytorch mobile QNNPACK.

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23658 Add support for using caffe2::ThreadPool in pytorch mobile QNNPACK.**

**How things work for caffe2:**
Caffe2 Ops -> NNPACK/QNNPACK -> pthreadpool_compute_1/2/3/4d_tiled -> pthreadpool_compute_1d (caffe2 shim) -> caffe2::ThreadPool

**Before this PR:**
Pytorch Ops -> NNPACK/QNNPACK -> pthreadpool_compute_1/2/3/4d_tiled -> pthreadpool_compute_1d (third_party implementation without mobile optimization)

caffe2::ThreadPool is optimized for mobile. This change leverages this logic for pytorch mobile as a temporary solution improve pytorch mobile perf. It is guarded by the C10_MOBILE macro.
For server side we return nullptr.

**Plan for next steps:**
Implement a mobile version of "at::parallel_for" which uses caffe2::ThreadPool internally so all ATen/TH multithreading usage is mobile optimized.
Refactor QNNPACK and/or pthreadpool to explicitly using "at::parallel_for" primitive to replace pthreadpool_compute_1d for Pytorch.
After QNNPACK is refactored, we will delete the mobile_threadpool() API.

Differential Revision: [D16594020](https://our.internmc.facebook.com/intern/diff/D16594020/)

supriyar committed:
Add support for using caffe2::ThreadPool in pytorch mobile QNNPACK.

caffe2::ThreadPool is optimized for mobile. This change leverages this logic for pytorch mobile. It is guarded by the C10_MOBILE macro.
For server side we return nullptr.

Differential Revision: [D16594020](https://our.internmc.facebook.com/intern/diff/D16594020/)

supriyar committed:
Update on "Add support for using caffe2::ThreadPool in pytorch mobile QNNPACK."


caffe2::ThreadPool is optimized for mobile. This change leverages this logic for pytorch mobile. It is guarded by the C10_MOBILE macro.
For server side we return nullptr.

Differential Revision: [D16594020](https://our.internmc.facebook.com/intern/diff/D16594020/)

ljk53 commented on aten/src/ATen/native/quantized/cpu/qnnpack_utils.h:
```diff
@@ -2,10 +2,26 @@
+#ifdef C10_MOBILE
+#include "caffe2/utils/threadpool/ThreadPool.h"
+#else
+#include <thread>
```
why do we need the <thread> here?

supriyar replied:
It is not, had something else in mind when I added it.

supriyar committed:
Update on "Add support for using caffe2::ThreadPool in pytorch mobile QNNPACK."


caffe2::ThreadPool is optimized for mobile. This change leverages this logic for pytorch mobile. It is guarded by the C10_MOBILE macro.
For server side we return nullptr.

Differential Revision: [D16594020](https://our.internmc.facebook.com/intern/diff/D16594020/)

supriyar committed:
Update on "Add support for using caffe2::ThreadPool in pytorch mobile QNNPACK."


caffe2::ThreadPool is optimized for mobile. This change leverages this logic for pytorch mobile. It is guarded by the C10_MOBILE macro.
For server side we return nullptr.

Differential Revision: [D16594020](https://our.internmc.facebook.com/intern/diff/D16594020/)

supriyar committed:
Update on "Add support for using caffe2::ThreadPool in pytorch mobile QNNPACK."


caffe2::ThreadPool is optimized for mobile. This change leverages this logic for pytorch mobile. It is guarded by the C10_MOBILE macro.
For server side we return nullptr.

Differential Revision: [D16594020](https://our.internmc.facebook.com/intern/diff/D16594020/)

ajtulloch commented on aten/src/ATen/native/quantized/cpu/qnnpack_utils.h:
```diff
@@ -2,10 +2,24 @@
+  static pthreadpool_t qnnpack_threadpool();
+#ifdef C10_MOBILE
+private:
+  static std::unique_ptr<caffe2::ThreadPool> thread_pool_;
```
Just FYI, because of the global lock `caffe2::ThreadPool::execution_mutex_`, this means two independent models executing convolutions on separate threads will contend on this shared resource. Not sure if this is a problem for you folks.

supriyar replied:
Thanks, I don't think this will affect our use case for now, but will keep it in mind. cc @ljk53

ljk53 replied:
Not sure if pytorch has something equivalent to c2's workspace - ideally these global resources should be allocated and owned by some global context object.

ajtulloch replied:
@ljk53 yes, the concept of "model" or "session" owned resources is a useful one that pytorch doesn't have I believe.

ljk53 replied:
cc @ilia-cher - have you thought about threading resource management API for devices? (for server it's probably fine as it's usually forever running binary anyway)

dzhulgakov replied:
As for "shared" thread pool - I'd say it's a good idea to have a single one. Is there any case we'd want to create thread pool per model?

Regarding clear up - it might be a good idea to clean up resources when the last Interpreter is destroyed assuming that there's no python extension present (though it still misses the case of C++ API just calling ops on tensors). Alternatively we can just expose a global call to "garbage collect resources" that tries to reset threading libraries.

What is the preferred strategy on device in general? if the last model is destroyed - clean up the thread pool?

ljk53 replied:
For the "shared" thread pool part - I agree it's bounded by cpu cores anyway unless we carefully config thread pool size for each model. But having a workspace kinda concept to hold all global resources is probably good for "clearing up" purpose? Right now I've seen static / thread-local scattered in the codebase. We might want to clean up at some point?

Yeah, we can associate resource clean up with the life-cycle of interpreter/predictor instance we expose to the application. My question is more about "at::parallel" level API to expose resource management functionality in an abstract way as "at::parallel" primitives hide thread-pool (or whatever implementation) as implementation details. Similarly - what if user wants to create and warm up thread-pool before first inference?

supriyar committed:
Update on "Add support for using caffe2::ThreadPool in pytorch mobile QNNPACK."


caffe2::ThreadPool is optimized for mobile. This change leverages this logic for pytorch mobile. It is guarded by the C10_MOBILE macro.
For server side we return nullptr.

Differential Revision: [D16594020](https://our.internmc.facebook.com/intern/diff/D16594020/)

ljk53 commented on aten/src/ATen/native/quantized/cpu/qnnpack_utils.cpp:
```diff
@@ -0,0 +1,20 @@
+#ifdef USE_QNNPACK
+#include <ATen/native/quantized/cpu/qnnpack_utils.h>
```
Shall we move the implementation to ATen/ParallelMobile.h (or some better name e.g. ParallelPThreadPool)? together with other ParallelOpenMP.h / ParallelNative.h / ParallelNativeTBB.h.

I think the reasons to not call it "qnnpack_xxx" are:
- NNPack on mobile should use this threadpool instead of 3rd-party/pthreadpool, right?
- We might want to provide parallel_for API (and other APIs defined in Parallel.h" for mobile. Probably should start with creating mobile friendly implementation there first, then hook up with the general parallel API.

Can we also follow the API convention there (use namespace instead of class + static methods)?

cc: @ilia-cher

supriyar committed:
Update on "Add support for using caffe2::ThreadPool in pytorch mobile QNNPACK."


caffe2::ThreadPool is optimized for mobile. This change leverages this logic for pytorch mobile. It is guarded by the C10_MOBILE macro.
For server side we return nullptr.

Differential Revision: [D16594020](https://our.internmc.facebook.com/intern/diff/D16594020/)

ljk53 commented on aten/src/ATen/ParallelMobileTPool.h:
```diff
@@ -0,0 +1,19 @@
+
+namespace at {
+namespace {
+static std::unique_ptr<caffe2::ThreadPool> thread_pool_ = nullptr;
```
I think you need move these to cpp files otherwise each compile unit will have their own instances...

ljk53 commented on aten/src/ATen/native/quantized/cpu/qnnpack_fc.cpp:
```diff
@@ -95,8 +96,13 @@ class QNNPACKLinear final : public torch::OperatorKernel {
         setupStatus == qnnp_status_success,
         "failed to setup QNNPACK Linear operator");
+    pthreadpool_t threadpool = nullptr;
+#ifdef C10_MOBILE
```
can we move this #ifdef to the implementation side instead? return null there if it's not mobile build.

ljk53 commented on aten/src/ATen/native/quantized/cpu/qnnpack_utils.h:
```diff
@@ -8,4 +8,5 @@ struct QnnpackOperatorDeleter {
     qnnp_delete_operator(op);
   }
 };
+
```
revert this?

supriyar committed:
Update on "Add support for using caffe2::ThreadPool in pytorch mobile QNNPACK."


caffe2::ThreadPool is optimized for mobile. This change leverages this logic for pytorch mobile. It is guarded by the C10_MOBILE macro.
For server side we return nullptr.

Differential Revision: [D16594020](https://our.internmc.facebook.com/intern/diff/D16594020/)

ljk53 commented on aten/src/ATen/ParallelMobileTPool.cpp:
```diff
@@ -0,0 +1,23 @@
+#include <ATen/ParallelMobileTPool.h>
+#include "caffe2/utils/threadpool/ThreadPool.h"
```
why you use '"' include here but '<' include above?

ljk53 commented on aten/src/ATen/ParallelMobileTPool.cpp:
```diff
@@ -0,0 +1,23 @@
+#include <ATen/ParallelMobileTPool.h>
```
nit: probably just call it ParallelMobile.h or ParallelMobileThreadPool.h?

ljk53 reviewed (commented):
Looks mostly good to me. I'll let @dzhulgakov  take the final pass.

supriyar committed:
Update on "Add support for using caffe2::ThreadPool in pytorch mobile QNNPACK."


caffe2::ThreadPool is optimized for mobile. This change leverages this logic for pytorch mobile. It is guarded by the C10_MOBILE macro.
For server side we return nullptr.

Differential Revision: [D16594020](https://our.internmc.facebook.com/intern/diff/D16594020/)

supriyar commented:
Addressed comments. I added some comments explaining that the usage of caffe2::Threadpool directly is temporary and that we will eventually have a mobile parallel_for implementation.

ilia-cher commented:
I really feel this is not compatible with we've been working on in H1, basically we don't want to have api such as "get thread pool" easily available - the whole point is to abstract away behind the primitives of ATen/Parallel.h and let parallel implementation handle what kind of thread pool is used, whether it is used and so on.

ilia-cher commented on aten/src/ATen/native/quantized/cpu/qnnpack_relu.cpp:
```diff
@@ -66,8 +67,10 @@ class QNNPACKRelu final : public torch::OperatorKernel {
+
     const qnnp_status runStatus =
-        qnnp_run_operator(qnnpack_operator, nullptr /* thread pool */);
+        qnnp_run_operator(qnnpack_operator, threadpool);
```
I believe a better approach would be to not pass thread pool as an explicit argument but have an operator itself call `at::parallel_for` and analogues, @supriyar have you checked what primitives are actually used in qnnpack implementations? We can add extra ones in ATen/Parallel.h and I can add a backend of ATen/Parallel.h specifically for mobile

ljk53 replied:
I totally agree that this is not compatible with the idea of aten::parallel - we should add more context to the PR description to give more context / future work plan.

- How things work for caffe2:
Caffe2 Ops -> NNPACK/QNNPACK -> pthreadpool_compute_1/2/3/4d_tiled -> pthreadpool_compute_1d (caffe2 shim) -> caffe2::ThreadPool

- Before this PR:
Pytorch Ops -> NNPACK/QNNPACK -> pthreadpool_compute_1/2/3/4d_tiled -> pthreadpool_compute_1d (third_party implementation without mobile optimization)

So our work plan is:
1. Take a baby step, make sure libtorch mobile links against caffe2/util/pthreadpool with caffe2::ThreadPool instead of the 3rd party one.

2. Implement a mobile version of "at::parallel_for" using caffe2::ThreadPool so all ATen/TH multithreading usage is mobile friendly;

3. Refactor QNNPACK or pthreadpool to explicitly using "at::parallel_for" primitive to replace pthreadpool_compute_1d for Pytorch;

When we get to step 3, we will delete the mobile_threadpool() API. This is just a temporary step to unblock us from running multi-thread QNNPack perf tests.

I feel the main reason to not pass thread-pool explicitly here is to hide threadpool as implementation detail. Whether to create threadpool upfront or lazily is somewhat orthogonal (see my comment about abstract resource management API above).

We could rename it to something like "_mobile_threadpool" or move it to a different util class without "ParallelXXX" prefix - but it's really just a temporary thing :)

dzhulgakov replied:
This explanation and plan makes sense, thanks! Definitely please copy paste in the PR description

supriyar replied:
Updated the PR description.

ilia-cher commented:
Instead of adding a public facing `at::mobile_threadpool()` can we move this new global (per-process) thread pool object into anonymous namespace within `utils/threadpool/pthreadpool.cc` and use it when threadpool pointer is nullptr (as of now in PT)?
This way you don't need to change PT ops to explicitly acquire thread pool (and pass it) and you also don't need to change QNNPACK and this shouldn't affect e.g. Caffe2 quantized ops that explicitly pass thread pool pointer.

ilia-cher commented on aten/src/ATen/ParallelMobileThreadPool.cpp:
```diff
@@ -0,0 +1,23 @@
+    thread_pool_ = caffe2::ThreadPool::defaultThreadPool();
+  }
+  return reinterpret_cast<pthreadpool_t>(thread_pool_.get());
+}
```
this thread pool is supposed to be created once and then accessed many times, we should think about an implementation where we don't have to take a lock each time we access the thread pool after it was created

ljk53 replied:
Considering the amount of work inside critical section (just read and cast a pointer) and the amount of work outside critical section (convolution), I'm not sure whether it's a premature optimization to use things like double checked locking and worry about memory barrier...

dzhulgakov commented on caffe2/utils/CMakeLists.txt:
```diff
@@ -19,7 +19,7 @@ list(APPEND Caffe2_CPU_SRCS
 # pthreadpool with a very similar interface. Neither NNPACK, nor this
 # thread pool supports Windows.
-if (NOT MSVC AND (NOT INTERN_BUILD_MOBILE OR BUILD_CAFFE2_MOBILE))
+if (NOT MSVC)
```
does it mean that we'd also link with this shim on server-side? it should be fine, but are there any special effects from it for the regular NNPACK?

ljk53 replied:
probably already linking this shim for server build - note that we haven't changed how things work for server build (prior to this change it was bypassing mobile-build unless it's old caffe2-mobile build)

supriyar committed:
Update on "Add support for using caffe2::ThreadPool in pytorch mobile QNNPACK."


caffe2::ThreadPool is optimized for mobile. This change leverages this logic for pytorch mobile. It is guarded by the C10_MOBILE macro.
For server side we return nullptr.

Differential Revision: [D16594020](https://our.internmc.facebook.com/intern/diff/D16594020/)

supriyar committed:
Update on "Add support for using caffe2::ThreadPool in pytorch mobile QNNPACK."


**How things work for caffe2:**
Caffe2 Ops -> NNPACK/QNNPACK -> pthreadpool_compute_1/2/3/4d_tiled -> pthreadpool_compute_1d (caffe2 shim) -> caffe2::ThreadPool

**Before this PR:**
Pytorch Ops -> NNPACK/QNNPACK -> pthreadpool_compute_1/2/3/4d_tiled -> pthreadpool_compute_1d (third_party implementation without mobile optimization)

caffe2::ThreadPool is optimized for mobile. This change leverages this logic for pytorch mobile as a temporary solution improve pytorch mobile perf. It is guarded by the C10_MOBILE macro.
For server side we return nullptr.

**Plan for next steps:**
Implement a mobile version of "at::parallel_for" which uses caffe2::ThreadPool internally so all ATen/TH multithreading usage is mobile optimized.
Refactor QNNPACK and/or pthreadpool to explicitly using "at::parallel_for" primitive to replace pthreadpool_compute_1d for Pytorch.
After QNNPACK is refactored, we will delete the mobile_threadpool() API.

Differential Revision: [D16594020](https://our.internmc.facebook.com/intern/diff/D16594020/)
